### PR TITLE
ERS DAI: Echo Reference Signal DAI and DMA implementation.

### PR DESCRIPTION
**Pull request by akloniex**
Implementation of virtual DAI and DMA dedicated for echo reference signal capture.

### CODE REVIEW

**Comment by jajanusz** [comment]
Needs #1635

**Comment by jajanusz** [comment]
@akloniex Assigning to ABI MINOR 9 atm

**Comment by tlauda** [comment]
@akloniex Could you describe how should the pipeline look like, what exactly does this dai do, why do we need additional DMA etc?

**Comment by akloniex** [comment]
@tlauda The main reason for creation of new DMA type is to provide continuous signal as echo reference, even when there is no playback running. In such case, ers-dma generates zeros at steady rate and pushed ers dai forward.
With previous implementation of echo reference, based on demux component, 2 pipelines were connected together, with echo pipe being dailess, which rendered it impossible to generate silence as output.

**Comment by tlauda** [comment]
@akloniex @mmaka1 @lbetlej 
I have concerns about adding the new DAI and DMA. DAI isn't really doing anything and DMA is "fake", but it's on the same list of platform's DMAs. Also it's not really flexible, because the naming suggests usage only for echo reference and the whole scheduling logic is inside DMA.
In my opinion adding additional "Virtual DAI" as audio component would be a better idea. The functionality would be the same. Scheduling would be set to timer, so no additional work here. Such component could be easily used in any other scenario e.g. probe with no additional implementations needed.

**Comment by akloniex** [comment]
@tlauda 
My intention was to make DAI reusable for probes and possibly other future features, but I ran short on names for things up my sleeve, so went the easy route of naming it as feature it was initially developed for. And naming it 'virtual' sounded very vague for me, so didn't decide to go with that name.
Nevertheless, if I understand correctly, your concern about DAI is mostly about it's name, not it's behaviour.
As for DMA, I believe that having it along with 'real' DMAs adds to consistency of the code, and in the long run should do more good than harm. Even though it doesn't represent any hardware behind it, from user point of view it behaves like one. Also, it shouldn't be able to choose by mistake, as it has unique supported device to connect to, which excludes it from dma_get() calls which don't request this type explicitly.

**Comment by tlauda** [comment]
@akloniex I haven't yet checked the DMA implementation itself. Wanted to start with the most basic thing, which is the structure and the location of the code. I fully understand the reasoning behind reusing DAI component and extracting the new functionality to DMA layer, but I just don't think it's the way to go in this particular case. There is nothing outstanding you need from the DAI's implementation and I would even say, that it adds complexity to your code, since you need to call different type of callbacks, take dma_sg_array etc.

File: src/include/user/trace.h [hunk 1]
```diff
@@ -53,11 +53,15 @@ struct system_time {
...
 #define TRACE_CLASS_CPU		(25 << 24)
+#define TRACE_CLASS_CLK		(26 << 24)
 #define TRACE_CLASS_EDF		(27 << 24)
 #define TRACE_CLASS_KPB		(28 << 24)
 #define TRACE_CLASS_SELECTOR	(29 << 24)
 #define TRACE_CLASS_SCHEDULE	(30 << 24)
 #define TRACE_CLASS_SCHEDULE_LL	(31 << 24)
+#define TRACE_CLASS_SOUNDWIRE	(32 << 24)
+#define TRACE_CLASS_KEYWORD	(33 << 24)
+#define TRACE_CLASS_ERS		(34 << 24)
```
Comment (lgirdwood) [inline src/include/user/trace.h]: This looks like it should be a separate patch.

File: src/platform/apollolake/include/platform/lib/dma.h [hunk 2]
```diff
@@ -22,6 +22,7 @@
 #define DMA_HOST_OUT_DMAC	5
 #define DMA_LINK_IN_DMAC	6
 #define DMA_LINK_OUT_DMAC	7
+#define DMA_ERS_DMAC		8
```
Comment (lgirdwood) [inline src/platform/apollolake/include/platform/lib/dma.h]: I'm really not following why this needs it's own DMAC ?

**Comment by jajanusz** [comment]
@akloniex Please resolve conflicta. Also there came at least 1 new dma ops that you need to add - `get_attribute`.

File: src/drivers/intel/cavs/ers.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+	if (!comp) {
+		trace_ers_error("ers_set_config() error: missing buffer "
+				"component to attach to, ID = %u",
+				config->ers.source_buffer_id);
+
+		return -ENODEV;
+	}
+
+	dai->plat_data.fifo[1].handshake = (uintptr_t)comp->cb;
```
[the remark this answers is not in the document]
Comment (akloniex) [reply src/drivers/intel/cavs/ers.c]: @lgirdwood comp in this context is ipc_comp_dev*, so comp->cb refers to comp_buffer pointer, not callback function. In this case buffer component is being passed as src_dev to dma, during dma_set_config call.
If you think that it's still too hacky, then I could try to find a better way to pass this pointer to dma.

[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/drivers/intel/cavs/ers.c]: yep, too hacky. please find a better way.

**Comment by lgirdwood** [comment]
@akloniex any update ?

**Comment by lbetlej** [comment]
@akloniex What are your plans regarding PR ? Do you agree to close it ?

**Comment by akloniex** [comment]
@lgirdwood @lbetlej I'm currently working on rebasing these changes and getting them back and running, but on a 2nd priority, after Probes. We could close this PR and open new one once I get it up, or we can keep it open and I'll just update it.

**Comment by lgirdwood** [comment]
@akloniex lets say v1.6 then. We can leave it open to keep it in ABI tracker

**Comment by jajanusz** [comment]
@akloniex Looks like stalled PR. Can I close it?

**Comment by paulstelian97** [comment]
> @akloniex Looks like stalled PR. Can I close it?

Activity in related PRs today (the optional Xrun things)
